build.zig.zon: module hash mismatch error is different on Windows

### Zig Version

0.11.0-dev.1580+a5b34a61a

### Steps to Reproduce and Observed Behavior

The test project [gettytestmodule.zip](https://github.com/ziglang/zig/files/10691521/gettytestmodule.zip) that was made with `zig init-exe` and not modified besides adding the `build.zig.zon` file behaves differently on 2 platforms:
1. on Windows x64, I think it calculates hashes wrong. However, when those are fixed then the `error: FileNotFound` output appears
2. on MacOS M1, the app just works when hashes are valid 

note: the zip project has "valid" hashes per Windows, not per Mac, I assume those are just invalid. 

The correct `build.zig.zon` should be:
```zig
.{
    .name = "app",
    .version = "1.0.0",
    .dependencies = .{
        .getty = .{
            .url = "https://github.com/getty-zig/getty/archive/f466b702f52895903248733ac0a304cc49d861f9.tar.gz",
            .hash = "1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed",
        },
        .json = .{
           .url = "https://github.com/getty-zig/json/archive/78fbee0b35d9b0bec7070e571cbe5e26bfbd1cce.tar.gz",
           .hash = "1220de8cd947effd84308509e4896639f5fe248a8a497761d0af3c45f04d94764b2a",
        },
    },
}
```

#### 1. STR: On Windows

1. Unpack the project
2. `zig build run`
Observe the output: `error: FileNotFound` despite the wrong module hashes.

3. remove the file `build.zig.zon` and run again `zig build run`
Observe the issue is gone. Project runs fine.

The module hashes in the  `build.zig.zon` file are actually wrong. If those are replaced with correct ones (calculated using MacOS M1) then observe the error about the hash mismatch:

```
w:\dev\gettytestmodule\build.zig.zon:7:21: error: hash mismatch: expected: 1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed, found: 1220748c2f35c4e2f213afacec95fa554082e891e356e995b5faf40eebca8dc4b82b
            .hash = "1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed",
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```


#### 2. STR: On MacOS M1

Running the project as is will result in the `error: hash mismatch` which is expected.
The module hashes in the  `build.zig.zon` file are wrong, so replace them with the correct ones. This will enable to run the project on MacOS.


### Expected Behavior

On Windows, the project build should:
1. not result in calculating wrong hashes on hash mismatch error
2. correctly build and run instead of resulting in `error: FileNotFound` when there is no hash mismatch error anymore

Looks like the issue is with https://github.com/ziglang/zig/blob/d3c9bfada64230ad17badb739f15a492b4c2c065/src/Package.zig#L630 On linux, `path` is something like `.github/ISSUE_TEMPLATE/bugs.yml`. On windows, `.github\ISSUE_TEMPLATE\bugs.yml`. Note the slashes.
### Proposed Solution
Add `fn normalizePath(path: []u8) void`. This can go in various places and do various things. Simplest is adding to `Package.zig` and just having it flip slashes. I guess that requires an alloc... Let me know if there are better ideas. I can implement this if y'all are cool with it.

Note: there's a fun new error on linux `0.11.0-dev.1783+436e99d13`
```
/home/ken/.cache/zig/p/1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed/build.zig:1:1: error: file exists in multiple modules
/home/ken/.cache/zig/p/1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed/build.zig:1:1: note: root of module root.@dependencies.json.getty
/home/ken/.cache/zig/p/1220f5207fb5410440310ead91178a430ea72293d4131ab0f990d75277723708d0ed/build.zig:1:1: note: root of module root.@dependencies.getty
```

--- a/lib/std/Build/CompileStep.zig
+++ b/lib/std/Build/CompileStep.zig
@@ -454,6 +454,7 @@ pub fn installConfigHeader(
         options.install_dir,
         dest_rel_path,
     );
+    install_file.step.dependOn(&config_header.step);
     cs.builder.getInstallStep().dependOn(&install_file.step);
     cs.installed_headers.append(&install_file.step) catch @panic("OOM");
 }

--- a/src/Package.zig
+++ b/src/Package.zig
@@ -493,6 +493,11 @@ fn fetchAndUnpack(
         // apply those rules directly to the filesystem right here. This ensures that files
         // not protected by the hash are not present on the file system.
 
+        // TODO: raise an error for files that have illegal paths on some operating systems.
+        // For example, on Linux a path with a backslash should raise an error here.
+        // Of course, if the ignore rules above omit the file from the package, then everything
+        // is fine and no error should be raised.
+
         break :a try computePackageHash(thread_pool, .{ .dir = tmp_directory.handle });
     };
 
@@ -546,7 +551,8 @@ fn unpackTarball(
 }
 
 const HashedFile = struct {
-    path: []const u8,
+    fs_path: []const u8,
+    normalized_path: []const u8,
     hash: [Manifest.Hash.digest_length]u8,
     failure: Error!void,
 
@@ -554,7 +560,7 @@ const HashedFile = struct {
 
     fn lessThan(context: void, lhs: *const HashedFile, rhs: *const HashedFile) bool {
         _ = context;
-        return mem.lessThan(u8, lhs.path, rhs.path);
+        return mem.lessThan(u8, lhs.normalized_path, rhs.normalized_path);
     }
 };
 
@@ -590,8 +596,10 @@ fn computePackageHash(
                 else => return error.IllegalFileTypeInPackage,
             }
             const hashed_file = try arena.create(HashedFile);
+            const fs_path = try arena.dupe(u8, entry.path);
             hashed_file.* = .{
-                .path = try arena.dupe(u8, entry.path),
+                .fs_path = fs_path,
+                .normalized_path = try normalizePath(arena, fs_path),
                 .hash = undefined, // to be populated by the worker
                 .failure = undefined, // to be populated by the worker
             };
@@ -609,12 +617,30 @@ fn computePackageHash(
     for (all_files.items) |hashed_file| {
         hashed_file.failure catch |err| {
             any_failures = true;
-            std.log.err("unable to hash '{s}': {s}", .{ hashed_file.path, @errorName(err) });
+            std.log.err("unable to hash '{s}': {s}", .{ hashed_file.fs_path, @errorName(err) });
         };
         hasher.update(&hashed_file.hash);
     }
     if (any_failures) return error.PackageHashUnavailable;
     return hasher.finalResult();
+}
+
+/// Make a file system path identical independently of operating system path inconsistencies.
+/// This converts backslashes into forward slashes.
+fn normalizePath(arena: Allocator, fs_path: []const u8) ![]const u8 {
+    const canonical_sep = '/';
+
+    if (fs.path.sep == canonical_sep)
+        return fs_path;
+
+    const normalized = try arena.dupe(u8, fs_path);
+    for (normalized) |*byte| {
+        switch (byte.*) {
+            fs.path.sep => byte.* = canonical_sep,
+            else => continue,
+        }
+    }
+    return normalized;
 }
 
 fn workerHashFile(dir: fs.Dir, hashed_file: *HashedFile, wg: *WaitGroup) void {
@@ -624,10 +650,10 @@ fn workerHashFile(dir: fs.Dir, hashed_file: *HashedFile, wg: *WaitGroup) void {
 
 fn hashFileFallible(dir: fs.Dir, hashed_file: *HashedFile) HashedFile.Error!void {
     var buf: [8000]u8 = undefined;
-    var file = try dir.openFile(hashed_file.path, .{});
+    var file = try dir.openFile(hashed_file.fs_path, .{});
     defer file.close();
     var hasher = Manifest.Hash.init(.{});
-    hasher.update(hashed_file.path);
+    hasher.update(hashed_file.normalized_path);
     hasher.update(&.{ 0, @boolToInt(try isExecutable(file)) });
     while (true) {
         const bytes_read = try file.read(&buf);

